File: python/flintdb/table.py
class Table:
    """Manages data within a specific table, handling row insertion updates, and lookups.

    Attributes:
        _name (str): The name of the table.
        _database (Database): The parent Database instance.
        _schema (list): The table defined scheme.
        _dek (str): The Data Encryption Key (DEK) for the table.
    """
    def __init__(self, name: str, database: "Database") -> None:
        """Initializes the Table object.

        Args:
            name: The name of the table.
            database: The parent Database instance.

        Raises:
            ValueError: If name contains invalid character.
            NotADirectoryError: If table does not exist.
        """
        self._name = name
        self._database = database
        self._schema = None
        self._dek = None

        if not self._name.isalnum():
            raise ValueError("Table name should contain only alphabets and numbers")
        elif not self.folder().is_dir():
            raise NotADirectoryError("Table does not exists")

    def name(self) -> str:
        """Returns table name."""
        return self._name

    def rename(self, new_name: str) -> bool:
        """Renames table to a desired new name.

        Args:
            new_name: The desired new name.

        Returns:
            True if successfully renamed, False otherwise.

        Raises:
            ValueError: If the new name contains invalid character.
        """
        if not new_name.isalnum():
            raise ValueError("Table name should contain only alphabets and numbers")

        target_dir = self._database.storage() / name_name
        if target_dir.is_dir():
            return False

        try:
            self.folder().replace(target_dir)
            from .cache import Cache
            cache = Cache(self)
            cache.flush()
            self._name = new_name
            return True
        except OSError:
            return False

    def folder(self) -> "Path":
        """Returns an instance of Path object, for the absolute path to the table directory"""
        return self._database.folder() / self._name

    def metadata(self, excess: bool = False) -> dict:
        """Returns table metadata.

        Args:
            excess: To get count of rows and total table size.
        """
        from .file import File

        folder = self.folder()
        file = File(folder / ".metadata")

        metadata = file.read_json()
        metadata["name"] = self._name
        metadata["modified"] = folder.stat().st_mtime
        metadata["rows"] = metadata["size"] = 0
        metadata["database"] = self._database

        self._dek = metadata["dek"]
        self._schema = metadata["schema"]

        if not excess:
            return metadata

        for row in self.rows():
            row_metadata = row.metadata()
            metadata["size"] += row_metadata["size"]
            metadata["rows"] += 1

        return metadata

    def delete(self) -> bool:
        """Deletes table directory."""
        try:
            import shutil
            from .cache import Cache
            shutil.rmtree(self.folder())
            cache = Cache(self)
            cache.flush()
            return True
        except OSError:
            return False

    def dek(self) -> str:
        """Returns table Data Encryption Key (DEK)."""
        if self._dek == None:
            self.metadata()

        return self._dek

    def database(self) -> "Database":
        """Returns parent database of the table."""
        return self._database

    def schema(self) -> dict:
        """Returns defined schema for the table."""
        if self._schema == None:
            self.metadata()

        return self._schema

    def alter(self, callback: "Callable") -> bool:
        """Updates table defined schema.

        Args:
            callback: The callback to return the instance of Schema object.

        Returns:
            True if successfully updated, False otherwise.

        Raises:
            TypeError: If callback returns something other than instance of Schema object.
        """
        from .schema import Schema
        try:
            schema = callback(Schema())
        except:
            schema = Schema()

        if not isinstance(schema, Schema):
            raise TypeError("Invalid schema provided")

        folder = self.folder()
        file = File(folder / ".metadata")

        schema.remove("_id")
        metadata = file.read_json()
        metadata["schema"] = schema.sorted_schema()
        return file.write_json(metadata)

    def insert(self, columns: dict) -> bool:
        """Inserts row to the table.

        Args:
            columns: The row columns to insert.

        Returns:
            True if successful, False otherwise.

        Raises:
            ValueError:
                If columns are empty.
                Or, if _id column is given and does not match any row.
                Or, if one of the columns has invalid value.
                Or, if one of the columns has encryption but database has no KEK.
                Or, if wrong database KEK is given.
        """
        if not columns:
            raise ValueError("Columns cannot be empty")

        folder = self.folder()
        if columns.get("_id"):
            try:
                self.row(columns.get("_id"))
            except ValueError:
                raise ValueError("ID does not match any row")
        else:
            from .crypto import Crypto
            id = Crypto.random_id(8)
            while (folder / f"{id}.ndjson").is_file():
                id = Crypto.random_id(8)

            columns["_id"] = id

        metadata = self.metadata()
        file = folder / f"{id}.ndjson"
        if file.is_file():
            row = self.row(columns["_id"])
            for key, value in row.columns():
                if key not in columns.keys():
                    columns[key] = value
        else:
            for column, options in metadata["schema"].items():
                if column not in columns.keys():
                    columns[column] = None

        has_encrypted_columns = False
        if metadata["schema"]:
            from .schema import Schema
            schema = Schema()
            for column, options in metadata["schema"].items():
                schema.add(column, options["type"], **options)

            has_encrypted_columns = schema.has_encrypted_columns()
            for key, value in columns.items():
                if not schema.valid(key, value):
                    raise ValueError(f"Invalid data type for column: {key}")

        del columns["_id"]
        columns = dict(sorted(columns.items()))

        import json
        data = json.dumps(list(columns.keys())) + "\n"
        if has_encrypted_columns:
            if not self._database.kek():
                raise ValueError("KEK required to encrypt columns")

            from .crypto import Crypto
            dek = Crypto.decrypt(self.dek(), self._database.kek())
            if not dek:
                raise ValueError("Wrong KEK provided")

            for key, value in columns.items():
                options = schema.get(key)
                if options and options["encrypted"]:
                    value = Crypto.encrypt(value, dek)

                data += json.dumps(value) + "\n"
        else:
            for key, value in columns.items():
                data += json.dumps(value) + "\n"

        from .file import File
        file = File(file)
        written = file.write(data)
        if written:
            from .cache import Cache
            cache = Cache(self)
            cache.flush()

        return written

    def insert_many(self, many_columns: list) -> list:
        """Insert rows to the table.

        Args:
            many_columms: The rows columns to insert.

        Returns:
            True for each successful insertion, and False for failed insertion.
        """
        status = []
        for columns in many_columns:
            status.append(self.insert(columns))

        return status

    def row(self, id: str) -> "Row":
        """Returns instance of Row object.

        Args:
            id: The row identifier.
        """
        from .row import Row
        return Row(id, self)

    def rows(self, exclude: list = []) -> "Generator":
        """Yields instances of Row object.

        Args:
            exclude: The rows to exclude.
        """
        for row in self.folder().iterdir():
            if row.name.endswith(".ndjson"):
                row = row.name.replace(".ndjson", "")
                if row not in exclude:
                    yield self.row(row)

    def query(self) -> "Query":
        """Returns an instance of the Query object for the table."""
        from .query import Query
        query = Query(self._database)
        query.use_table(self._name)
        return query

    def find(self, criteria: dict) -> list:
        """Returns matching rows from the table.

        Args:
            criteria: The condition to match.
        """
        query = self.query()
        for key, value in criteria.items():
            query.where(key, "=", value)

        return query.fetch()

    def find_one(self, criteria: dict) -> "Any":
        """Returns single matching row from the table.

        Args:
            criteria: The condition to match.
        """
        query = self.query()
        for key, value in criteria.items():
            query.where(key, "=", value)

        query.limit(1)
        query.no_cache()
        data = query.fetch()
        if data:
            return data[0]

        return None

File: php/flintdb/Table.php
<?php

namespace FlintDB;

use DirectoryIterator,
    DomainException,
    Generator,
    LogicException,
    RuntimeException;

class Table {
  /**
   * The name of the table
   *
   * @var string
   */
  private string $name;
  
  /**
   * The table defined schema
   *
   * @var array|null
   */
  private ?array $schema = null;
  
  /**
   * The table data encryption key
   *
   * @var string|null
   */
  private ?string $dek = null;
  
  /**
   * The parent database name
   *
   * @var string
   */
  private string $database;
  
  /**
   * The parent database directory
   *
   * @var string
   */
  private string $storage;
  
  /**
   * The parent database encryption key
   *
   * @var string
   */
  private string $kek;
  
  /**
   * Constructor
   *
   * @param string $name
   * @param string $database
   * @param string $storage
   * @param string $kek
   * @throws DomainException
   * @throws RuntimeException
   */
  public function __construct( string $name, string &$database, string &$storage, #[\SensitiveParameter] string &$kek ) {
    $this->name = $name;
    $this->database =& $database;
    $this->storage =& $storage;
    $this->kek =& $kek;
    
    if ( ! ctype_alnum( $this->name ) ) {
      throw new DomainException( 'Table name should contain only alphabets and numbers' );
    }
    
    elseif ( ! is_dir( $this->folder() ) ) {
      throw new RuntimeException( 'Table does not exist' );
    }
  }
  
  /**
   * Returns table name
   *
   * @return string
   */
  public function &name(): string {
    return $this->name;
  }
  
  /**
   * Renames table
   *
   * @param string $new_name
   * @return string
   * @throws DomainException
   */
  private function rename(): bool {
    if ( ! ctype_alnum( $new_name ) ) {
      throw new DomainException( 'Table name should contain only alphabets and numbers' );
    }
    
    $target = Filesystem::join( $this->storage, $this->database, $this->name );
    return rename( $this->folder(), $target );
  }
  
  /**
   * Returns table directory path
   *
   * @return string
   */
  public function folder(): string {
    return Filesystem::join( $this->storage, $this->database, $this->name );
  }
  
  /**
   * Returns table metadata
   *
   * @param bool $excess
   * @return array
   */
  public function metadata( bool $excess = false ): array {
    $folder = $this->folder();
    $file = Filesystem::join( $folder, '.metadata' );
    $metadata = Filesystem::read_json( $file );
    $metadata[ 'modified' ] = filemtime( $folder );
    $metadata[ 'rows' ] = $metadata[ 'size' ] = 0;
    $metadata[ 'database' ] = $this->database;
    $metadata[ 'name' ] = $this->name;
    
    $this->dek = $metadata[ 'dek' ];
    $this->schema = $metadata[ 'schema' ];
    
    if ( ! $excess ) {
      return $metadata;
    }
    
    foreach ( $this->rows() as $row ) {
      $metadata[ 'size' ] += $row->metadata()[ 'size' ];
      $metadata[ 'rows' ]++;
    }
    
    return $metadata;
  }
  
  /**
   * Deletes table directory
   *
   * @return bool
   */
  public function delete(): bool {
    $target = Filesystem::join( $this->storage, $this->database, '.deleted_' . $this->name );
    if ( ! rename( $this->folder(), $target ) ) {
      return false;
    }
    
    Filesystem::rmtree( $target );
    $cache = new Cache( $this );
    $cache->flush();
    return true;
  }
  
  /**
   * Returns table Data Encryption Key (DEK)
   *
   * @return string
   */
  public function &dek(): string {
    if ( is_null( $this->dek ) ) {
      $this->metadata();
    }
    
    return $this->dek;
  }
  
  /**
   * Returns table schema
   *
   * @return array
   */
  public function &schema(): array {
    if ( is_null( $this->schema ) ) {
      $this->metadata();
    }
    
    return $this->schema;
  }
  
  /**
   * Modifies table schema
   *
   * @param callable $callback
   * @return bool
   * @throws LogicException
   */
  public function alter( callable $callback ): bool {
    try {
      $schema = $callback( new Schema );
    } catch ( Throwable $err ) {
      $schema = new Schema;
    }
    
    if ( ! $schema instanceof Schema ) {
      throw new LogicException( 'Invalid schema provided' );
    }
    
    $schema->remove( '_id' );
    
    $metadata = array();
    $metadata[ 'schema' ] = $schema->sorted_schema();
    $file = Filesystem::join( $this->folder(), '.metadata' );
    return Filesystem::write_json( $file, $metadata );
  }
  
  /**
   * Inserts row to the table
   *
   * @param array $columns
   * @return bool
   * @throws RuntimeException
   * @throws DomainException
   * @throws LogicException
   */
  public function insert( array $columns ): bool {
    $folder = $this->folder();
    if ( isset( $columns[ '_id' ] ) ) {
      try {
        $this->row( $columns[ '_id' ] );
      } catch ( RuntimeException $err ) {
        throw new RuntimeException( 'ID does not match any row' );
      }
    } else {
      $id = Crypto::random_id(8);
      while ( is_file( $folder . '/' . $id . '.ndjson' ) ) {
        $id = Crypto::random_id(8);
      }
      
      $columns[ '_id' ] = $id;
    }
    
    $metadata = $this->metadata();
    $file = Filesystem::join( $folder, $columns[ '_id' ] . '.ndjson' );
    if ( is_file( $file ) ) {
      $row = $this->row( $columns[ '_id' ] );
      foreach ( $row->yield_columns() as $column => $value ) {
        if ( ! isset( $columns[ $column ] ) ) {
          $columns[ $column ] = $value;
        }
      }
    } else {
      foreach ( $metadata[ 'schema' ] as $column => $options ) {
        if ( ! isset( $columns[ $column ] ) ) {
          $columns[ $column ] = null;
        }
      }
    }
    
    $schema = new Schema( $metadata[ 'schema' ] );
    foreach ( $columns as $column => $value ) {
      if ( ! $schema->valid( $column, $value ) ) {
        throw new DomainException( 'Invalid data type for column: ' . $column );
      }
    }
    
    unset( $columns[ '_id' ] );
    ksort( $columns );
    
    $data = json_encode( array_keys( $columns ) ) . PHP_EOL;
    if ( $schema->has_encrypted_columns() ) {
      if ( empty( $this->kek ) ) {
        throw new LogicException( 'KEK required to encrypt columns' );
      }
      
      $dek = Crypto::decrypt( $metadata[ 'dek' ], $this->kek );
      if ( ! $dek ) {
        throw new LogicException( 'Invalid KEK provided' );
      }
      
      foreach ( $columns as $column => $value ) {
        $column = $schema->get( $column );
        if ( $column[ 'encrypted' ] ) {
          $value = Crypto::encrypt( $value, $dek );
        }
        
        $data .= json_encode( $value ) . PHP_EOL;
      }
    }
    
    else {
      foreach ( $columns as $column => $value ) {
        $data .= json_encode( $value ) . PHP_EOL;
      }
    }
    
    $written = Filesystem::write( $file, $data );
    if ( $written ) {
      $cache = new Cache( $this );
      $cache->flush();
    }
    
    return $written;
  }
  
  /**
   * Insert many rows to the table
   *
   * @param array[] $many_columns
   * @return array
   */
  public function insert_many( array ...$many_columns ): array {
    $status = array();
    foreach ( $many_columns as $index => $columns ) {
      $status[ $index ] = $this->insert( $columns );
    }
    
    return $status;
  }
  
  /**
   * Returns instance of Row class
   *
   * @param string $id
   * @return Row
   */
  public function row( string $id ): Row {
    return new Row( $id, $this->name, $this->schema(), $this->dek(), $this->database, $this->storage, $this->kek );
  }
  
  /**
   * Yields all rows
   *
   * @param array $exclude
   * @return Generator
   */
  public function rows( array $exclude = [] ): Generator {
    foreach ( new DirectoryIterator( $this->folder() ) as $row ) {
      if ( str_ends_with( (string) $row, '.ndjson' ) ) {
        $row = $row->getBasename( '.ndjson' );
        if ( ! in_array( $row, $exclude ) ) {
          yield $this->row( $row );
        }
      }
    }
  }
  
  /**
   * Returns instance of Query class
   *
   * @return Query
   */
  public function query(): Query {
    $database = new Database( $this->database, $this->storage, $this->kek );
    return ( new Query( $database ) )->from( $this->name );
  }
  
  /**
   * Returns matching rows from the table
   *
   * @param array $criteria
   * @return Collection
   */
  public function find( array $criteria ): Collection {
    $query = $this->query();
    foreach ( $criteria as $key => $value ) {
      $query->where( $key, '=', $value );
    }
    
    $query->no_cache();
    return $query->fetch();
  }
  
  /**
   * Returns matching rows from the table
   *
   * @param string $id
   * @return ?Row
   */
  public function find_one( array $criteria ): ?Row {
    $query = $this->query();
    foreach ( $criteria as $key => $value ) {
      $query->where( $key, '=', $value );
    }
    
    $query->limit(1);
    $query->no_cache();
    $data = $query->fetch();
    return $data->item(0);
  }
}
